src/Toolkit/Toolkit/UI/Controls/Legend/LegendDataSource.cs

```diff
@@ -0,0 +1,541 @@
+                    i--;
+                }
+                else if (!_items.Contains(newItem))
+                {
+                    // Item was added
+                    _items.Insert(i, newItem);
+                    OnCollectionChanged(new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Add, newItem, i));
+                }
+                else
+                {
+                    // Item was removed (or moved)
+                    var removedItem = oldItem;
+                    _items.RemoveAt(i);
+                    OnCollectionChanged(new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Remove, oldItem, i));
+                    i--;
+                }
+            }
+#if DEBUG
+            // Validate the calculated collection is in sync
+            System.Diagnostics.Debug.Assert(newItems.Count == _items.Count, "Legend entry count doesn't match");
+            for (i = 0; i < newItems.Count; i++)
+            {
+                System.Diagnostics.Debug.Assert(newItems[i] == _items[i], $"Legend entry {i} doesn't match");
+            }
+#endif
+            _items = newItems;
+        }
+
+        private void GeoView_LayerViewStateChanged(object sender, LayerViewStateChangedEventArgs e) => MarkCollectionDirty();
+
+        private void GeoView_PropertyChanged(object sender, PropertyChangedEventArgs e)
+        {
+#if XAMARIN || XAMARIN_FORMS
+            if ((sender is MapView && e.PropertyName == nameof(MapView.Map)) ||
+                (sender is SceneView && e.PropertyName == nameof(SceneView.Scene)))
+            {
+                UpdateItemsSource();
+            }
+#endif
+            if (e.PropertyName == nameof(MapView.MapScale) && sender is MapView mv)
+            {
+                _currentScale = mv.MapScale;
+                MarkCollectionDirty();
+            }
+        }
+
+        private double _currentScale = double.NaN;
+
+        private List<object> BuildLegendList(IEnumerable<ILayerContent> layers)
+        {
+            if (layers == null)
+            {
+                return null;
+            }
+
+            List<object> data = new List<object>();
+            foreach (var layer in layers)
+            {
+                if (!layer.ShowInLegend || (!layer.IsVisible && _filterHiddenLayers))
+                {
+                    continue;
+                }
+
+                if (layer is Layer l)
+                {
+                    var state = _geoview.GetLayerViewState(l);
+                    if (state != null &&
+                        ((state.Status == LayerViewStatus.NotVisible && _filterHiddenLayers) ||
+                        (state.Status == LayerViewStatus.OutOfScale && _filterByVisibleScaleRange)))
+                    {
+                        continue;
+                    }
+                }
+                else if (layer is ILayerContent ilc)
+                {
+                    if (!ilc.IsVisible && _filterHiddenLayers)
```

Comment:
Note: WMSSublayers are hierarchical, and parent sublayers do NOT affect visibility of child sublayers.